File: include/img/MILTrack/matlab/用haar特征/my_imadjust.m
function OutImage = my_imadjust(InImage)
%函数功能：增加灰度图像的对比度，等效于matlab自带函数imadjust,即等效于OutImage = imadjust(InImage);
%InImage为输入的灰度图像
%OutImage为增加对比度后的灰度图像

%以下代码和MATLAB自带函数stretchlim等同，即等效于lowhigh_in = stretchlim(InImage);%
nbins = 256;
tol = [.01 .99];
tol_low = tol(1);
tol_high = tol(2);
N = imhist(InImage,nbins);%求256个bin的直方图
cdf = cumsum(N)/sum(N);%cumsum(N)为求N中元素的累加和
low = find(cdf>tol_low, 1, 'first');%此句意思是返回cdf中的第一个大于tol_low的元素索引
high = find(cdf>=tol_high, 1, 'first');
if low==high
    lowhigh(:,1) = [1; nbins];
else
    lowhigh(:,1) = [low;high];
end
lowhigh_in = (lowhigh - 1)/(nbins-1);  % convert to range [0 1]
%上代码和MATLAB自带函数lowhigh_in = stretchlim(InImage);等同%


low_in   = lowhigh_in(1);
high_in  = lowhigh_in(2);
low_out  = 0;
high_out = 1;

temp1 = linspace(0,1,256);
temp1(:) = max(low_in,min(high_in,temp1));

temp2 = ((temp1-low_in)./(high_in-low_in));
temp2(:) = temp2.*(high_out-low_out) + low_out;

lut = im2uint8(temp2);%查找表所用到的表，即lut = uint8(round(temp2.*255));
OutImage = zeros(size(InImage,1),size(InImage,2));
OutImage= uintlut(InImage,lut);%如InImage中第100个位置的的元素值为a,则OutImage相同位置的元素值为lut(a+1)

File: include/img/MILTrack/my_imadjust.c

#include <math.h>
#include <stdio.h>
#include "_MILTrack.h"
#include "direct.h"
#include "utime.h"

// 用映射表m把f 映射到 p
#define MAPPING(_N, _A, _B, _M) {int _I=0; for (; _I<_N; ++_I) { _B[_I] = _M[(int)(_A[_I])]; } }

#define HIST(_N, _A, _H) {int _I=0; for (; _I<_N; ++_I) { ++_H[(int)(_A[_I])]; } }

void histeq( int _N, const unsigned char* f, unsigned char* p ) {
  enum { BIT = 8, SCAL = 14, };
  enum { LV = 1 << BIT, };
  const int MN = ( LV << SCAL ) / ( _N );
  int t, i; // 循环变量
  unsigned char bMap[ LV ] = {0}; // 灰度映射表
  int nCnt[ LV ] = {0}; // 累积直方图
  HIST( _N, f, nCnt );
  bMap[ 0 ] = ( nCnt[ 0 ] * MN );
  for ( i = 1; i < LV; ++i ) {
    nCnt[ i ] += nCnt[ i - 1 ];
    t = ( nCnt[ i ] * MN ) >> SCAL;
    bMap[ i ] = ( t >= LV ) ? ( LV - 1 ) : ( t ); // 计算对应的新灰度值
  }
  MAPPING( _N, f, p, bMap );
  return ;
}

//函数功能：增加灰度图像的对比度，等效于matlab自带函数imadjust,即等效于OutImage = imadjust(InImage);
//InImage为输入灰度的图像
//OutImage为增加对比度后的灰度图像
void my_imadjust( int _N, const unsigned char* f, unsigned char* p ) {
  enum { BIT = 8, SCAL = 14, };
  enum { LV = 1 << BIT, };
  int t, i; // 循环变量
  unsigned char bMap[ LV ] = {0}; // 灰度映射表
  int nCnt[ LV ] = {0}; // 累积直方图
  int low_in;
  int high_in;
  int tol_low = ( int ) ( 0.01 * _N );
  int tol_high = ( int ) ( 0.01 * _N );
  HIST( _N, f, nCnt );
  for ( t = 0, low_in = 0; low_in < LV; ++low_in ) {
    t += nCnt[ low_in ];
    if ( t > tol_low ) {
      break;
    }
  }
  for ( t = 0, high_in = LV; high_in-- > 0; ) {
    t += nCnt[ high_in ];
    if ( t > tol_high ) {
      break;
    }
  }
  {
    int len = ( high_in - low_in );
    for ( i = 0; i < LV; ++i ) {
      if ( i < low_in ) {
        bMap[ i ] = 0;
      } else if ( i >= high_in ) {
        bMap[ i ] = LV - 1;
      } else {
        //查找表所用到的表，即lut = uint8(round(temp2.*255));
        bMap[ i ] = ( LV * ( i - low_in ) ) / len; // 计算对应的新灰度值
      }
    }
  }
  //如InImage中第100个位置的元素值为a,则OutImage相同位置的元素值为lut(a+1)
  MAPPING( _N, f, p, bMap );
  return ;
}
